text/0000-disconnect-flow.md

```diff
@@ -0,0 +1,134 @@
+
+When a `DISCONNECT` message is received:
+
+- If the recipient is a client and the clientId is its own `LocalClientId` or the `ServerClientId`, it calls the `OnServerConnectionLostCallback` with the provided reason and `NetworkManager.StopClient()` or `NetworkManager.StopHost()` is called automatically as appropriate. **This is a change: The client will no longe receive the OnClientDisconnected callback, which is now server-only**
+
+- If the recipient is a client and a clientId other than `LocalClientId` or `ServerClientId` is provided, it calls the `OnPeerDisconnectedCallback` with the provided reason.
+
+- If the recipient is a server, it calls the `OnClientDisconnectedCallback` with `DisconnectReason.Stopped`
+
+Disconnect callbacks can also be invoked in the following circumstances:
+
+- When the server detects a timeout on a client connection, it will call the `OnClientDisconnectedCallback` with `DisconnectReason.TimedOut`
+
+- When the client detects a timeout on the server connection, it will call the `OnServerConnectionLostCallback` with `DisconnectReason.TimedOut`
+
+- When a server closes an established client's connection for any reason, it will call the `OnClientDisconnectedCallback` with `DisconnectReason.Kicked`
+
+**Note: ConnectedClients and ConnectedClientsList do not get updated, as clients cannot talk directly to each other so there is no NetworkClient to provide.**
+
+# Drawbacks
+
+Other than adding a nearly negligible amount of network traffic, there are no drawbacks to speak of.
+
+# Rationale and alternatives
+
+Providing this information to clients allows clients to more intelligently react to connection changes. In particular, the `DisconnectReason` field is useful for messaging about events to a player - while the code may handle all types of disconnects the same, it is useful to be able to communicate to the player the difference between a network timeout, a server shutdown, an authorization failure, and a kick.
+
+# Prior art
+
+TODO
+
+- 
+
+# Unresolved questions
+
+- Knowing that we can't update ConnectedClients and ConnectedClientsList, should we have a `HashSet<ulong> PeerList` available to users so they don't have to maintain such a list themselves?
+- Is there any possibility that providing a client with a list of anonymous client IDs for other clients poses a security risk? It's not clear what they could do with that information that would pose such a risk, especially since any objects owned by other clients will expose that same client ID.
+
+# Future possibilities
+
+There is the possibility additional disconnect reasons may be added in the future, or even the possibility we may want to provide the ability for the user to provide their own disconnect reasons. A `NetworkManager.Kick(clientId, reason)` method seems valuable. In that situation, we may consider making the reason a ulong with reserved values so users can add additional disconnect codes, or possibly even a string.
```

Comment:
It'd be awesome if we could have the ability to provide a payload with our responses.
Right now in Boss Room, we have this very ugly workaround with a custom message to give clients a reason for disconnection.
https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/9960a59586a521cf15cb8b477b2aa1c7370f3e13/Assets/BossRoom/Scripts/Server/Net/ServerGameNetPortal.cs#L292
https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/9960a59586a521cf15cb8b477b2aa1c7370f3e13/Assets/BossRoom/Scripts/Shared/Net/GameNetPortal.cs#L255

Do you think this future possibility will be in for our fall release?

Reply:
I don't think it'd be difficult to add this for the fall release. I was trying to keep this RFC a little slimmer and not try to spec out the ways a client could get willfully disconnected from the server, but if it's valuable to add now, I'm on board with adding that functionality to this RFC. I can't speak to scheduling of implementation though, of course. :)

I'll update the RFC to add the ability for the disconnect to carry a custom UserData buffer that can be passed opaquely into the callback. Do you think that would solve your need here? Something like `void Kick(ulong clientId, DisconnectReason reason, NetworkBuffer userDataBuffer = null)` - or is there a way that the third parameter could be an `INetworkSerializable` and get properly constructed on the other end without the user having to deal with buffers directly? (I'm not sure if INetworkSerializables pass any type information with them that could be used to construct it on the other side in a generic way or not.)

Reply:
ok ok, it's really up to you. I just wanted to make sure this didn't fall into cracks :)
we have multiple ways to handle that serialization (that we really should make unique) You can look into how netvars currently do it. Thing is, there's plans with snapshot to look at serialization, you might want to have a quick sync with @jeffreyrainy and @LukeStampfli 
Do we need type information though? Both client and server would be implemented by the same team, the person implementing the client would know what data type the server sends in their own custom payload and would be able to deserialize using the appropriate type, no?

Reply:
The question I had here was, could we make something like

```OnPeerDisconnectedCallback(DisconnectReason reason, INetworkSerializable payload)```

And have `payload` get deserialized 'magically' by the SDK code, which wouldn't know what type `payload` is expected to be? I'm guessing probably not - which is why I proposed using something like `NetworkBuffer`, but I do feel that `NetworkBuffer` is a bit more inaccessible to less experienced teams, so if there was a way to make the payload magically deserialize to the correct type, it'd be a better developer experience.

In this case, unlike RPCs and NetVars, we don't necessarily have any shared understanding on both sides at compile time of what that type will be, due to the generic nature of the callback type signature.

Reply:
We currently do not support typed serialization and I don't think it is a feature which we can just quickly add to MLAPI.

What we could do is use generics for this. The API would then look like the following:
```csharp
OnPeerDisconnectedCallback(DisconnectReason reason, Payload payload)
```

`Payload` would be a struct which a) exposes a `NetworkBuffer` property for direct network buffer access and b) it will have a `As<T>` function to convert the payload automagically into the right type. So for instance:
`payload.As<MyCustomSerializable>();` would extract the payload as a `MyCustomSerializable`.

The only restriction here would be that the sides writing and reading the payload must use the same generic type. But I'd argue that is something that we want because it is much cleaner then typed serialization.

Reply:
100% agree, typed serialization is a little scary. Generics is a better option, gets the ability to use INetworkSerializables so it's accessible but it doesn't add the bandwidth and CPU overhead of sending and decoding type info.